direnv.el

```diff
@@ -102,42 +103,73 @@ use `default-directory', since there is no file name (or directory)."
+called. If direnv produced no output at all, nothing is done: no
+CALLBACK, no user message, no error."
   (unless direnv--executable
     (setq direnv--executable (direnv--detect)))
-  (let ((environment process-environment)
-        (stderr-tempfile (make-temp-file "direnv-stderr"))) ;; call-process needs a file for stderr output
-    (unwind-protect
-        (with-current-buffer (get-buffer-create direnv--output-buffer-name)
-          (erase-buffer)
-          (let* ((default-directory directory)
-                 (process-environment environment)
-                 (exit-code (call-process
-                             direnv--executable nil
-                             `(t ,stderr-tempfile) nil
-                             "export" "json")))
-            (prog1
-                (unless (zerop (buffer-size))
-                  (goto-char (point-max))
-                  (re-search-backward "^{")
-                  (let ((json-key-type 'string)
-                        (json-object-type 'alist))
-                    (json-read-object)))
-              (unless (zerop (direnv--file-size stderr-tempfile))
-                (goto-char (point-max))
-                (unless (zerop (buffer-size))
-                  (insert "\n\n"))
-                (insert-file-contents stderr-tempfile))
-              (with-temp-buffer
-                (unless (zerop exit-code)
-                  (insert-file-contents stderr-tempfile)
-                  (display-warning
-                   'direnv
-                   (format-message
-                    "Error running direnv (exit code %d):\n%s\nOpen buffer ‘%s’ for full output."
-                    exit-code (buffer-string) direnv--output-buffer-name)))))))
-      (delete-file stderr-tempfile))))
+  (let (;; Cannot use temp buffer because the lexical scope ends before the
+        ;; async lifetime, causing the temporary buffer to be cleaned up before
+        ;; the process is done. By using a deterministic name here we ensure
+        ;; only one instance of direnv runs at a time. This is suboptimal: it
+        ;; would be nice to have parallel direnvs for every separate directory,
+        ;; but that would require normalising which direnv roots for buffers in
+        ;; different subdirectories of the same project, which can get tricky,
+        ;; while the benefit isn’t that pronounced (just a little wait
+        ;; longer). Meanwhile, the benefit of not having a parallel direnv for
+        ;; the same directory is very clear, particularly for long-running calls
+        ;; e.g. the first time you download a nix tree.
+        (stdout-buffer " *direnv stdout*")
+        (environment process-environment))
+    (if (buffer-live-p (get-buffer stdout-buffer))
+        (progn
+          (message "direnv: already running, retrying %s in 1 second (see %s)"
+                   directory direnv--output-buffer-name)
+          (run-with-timer 1 nil 'direnv--export directory callback t))
```

Comment:
Do you mean https://www.gnu.org/software/emacs/manual/html_node/elisp/Transaction-Queues.html ?

Reply:
No, sorry, I just mean a list of "things to do next". In this case, it could just be a list of `(directory . callback)` pairs.

The current code is using timers to poll for when they are able to run. This is nondeterministic (i.e. the tasks may run out of order), and uses cycles polling. These are unnecessary downsides, though, because you know when you're finished because the `on-success` callback below runs.

Instead, you could `(push (list directory callback) direnv--export-queue)` instead of setting a timer, then `(apply #'direnv--export (pop direnv--export-queue))` at the end of `on-success`, `on-error`, and the `error` handler in `condition-case`.